### PR TITLE
Remove vulcan-results dependency

### PR DESCRIPTION
**Pull request by danfaizer**
This PR eliminates the dependency on vulcan-results.

Currently, vulcan-agent runs vulcan-checks, and then sends the resulting vulnerability report and raw logs to vulcan-results for storage. However, this process can be made more efficient by having vulcan-agent handle the file uploads directly, instead of relying on vulcan-results.

By removing this extra layer, we can avoid unnecessary roundtrips between services and improve scalability. This change means that the capacity for file uploads will automatically adjust to match the number of vulcan-agents, which is a more optimal solution in the long term.
Additionally, this modification has the potential to reduce costs by minimizing cpu and memory usage for vulcan-results and decreasing overall bandwidth consumption.

Merging this PR requires two previous changes:

1. Add S3 write permissions on the target buckets to the vulcan-agent IAM role.
2. Modify the vulcan-agent config to set up the new required parameters:
```
[s3_writer]
link_base = "http://vulcan-results.example.com/v1/"
bucket_reports = "my-vulcan-check-reports"
bucket_logs = "my-vulcan-check-logs"
region = "eu-west-1"
s3_link = false
```

These changes are necessary to enable the vulcan-agent to directly upload the reports and logs to the specified S3 buckets without relying on vulcan-results.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -35,13 +35,12 @@ import (
 // 0 if the agent terminated gracefully, either by receiving a TERM signal or
 // because it passed more time than configured without reading a message.
 func Run(cfg config.Config, b backend.Backend, l log.Logger) int {
```
Comment (jesusfcr) [inline agent/agent.go:37]: I would move the Uploader config outside the Run

- Define Uploader as a simple interface with the send methods.
- adding a new parameter uploader that implements that `Uploader` interface
- Create the S3Uploader that implements this Uploader interface and is configured in the main.go.

With this simple trick will allow i.e. `lava` to provide a trivial implementation to receive the data.

File: storage/s3/writer.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		bucket = w.cfg.BucketReports
+	case "logs":
+		extension = "log"
+		contentType = "text/plain"
+		bucket = w.cfg.BucketLogs
+	default:
+		return "", ErrUnsupportedKind
+	}
+
+	key = fmt.Sprintf("%s/%s.%s", kind, checkID, extension)
```
Comment (julianvilas) [inline storage/s3/writer.go:99]: Just out of curiosity: we are not using anymore the date partitioning schema for the athena queries, right?

File: storage/s3/writer_test.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		bucket        string
+		s3link        bool
+		expectedLink  string
+		expectedError error
+	}{
+		{
+			name:          "ValidKindReports",
+			kind:          "reports",
+			extension:     "json",
+			contentType:   "test/json",
```
Comment (julianvilas) [inline storage/s3/writer_test.go:144]: Shouldn't this be `text` instead?

File: storage/s3/writer_test.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+			bucket:        "bucketReports",
+			s3link:        false,
+			expectedLink:  "https://www.example.com/reports/dt=2023-05-12/check123/check123.json",
+			expectedError: nil,
+		},
+		{
+			name:          "ValidKindLogs",
+			kind:          "logs",
+			extension:     "log",
+			contentType:   "test/plain",
```
Comment (julianvilas) [inline storage/s3/writer_test.go:154]: Shouldn't this be `text` instead?

File: storage/s3/writer_test.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+			bucket:        "bucketLogs",
+			s3link:        false,
+			expectedLink:  "https://www.example.com/logs/dt=2023-05-12/check123/check123.log",
+			expectedError: nil,
+		},
+		{
+			name:          "ValidKindReportsS3Link",
+			kind:          "reports",
+			extension:     "json",
+			contentType:   "test/json",
```
Comment (julianvilas) [inline storage/s3/writer_test.go:164]: Shouldn't this be `text` instead?

File: storage/s3/writer_test.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+			bucket:        "bucketReports",
+			s3link:        true,
+			expectedLink:  "s3://bucketReports/reports/check123.json",
+			expectedError: nil,
+		},
+		{
+			name:          "ValidKindLogsS3Link",
+			kind:          "logs",
+			extension:     "log",
+			contentType:   "test/plain",
```
Comment (julianvilas) [inline storage/s3/writer_test.go:174]: Shouldn't this be `text` instead?

**Review by julianvilas — CHANGES_REQUESTED** [review]
Very good job mate! Just some minor comments.

File: storage/s3/writer.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+)
+
+var (
+	ErrReportsBucketNotDefined = errors.New("reports bucket must be defined")
+	ErrLogsBucketNotDefined    = errors.New("logs bucket must be defined")
+	ErrLinkBaseNotDefined      = errors.New("link base must be defined or s3 link enabled")
+	ErrUnsupportedKind         = errors.New("unsupported kind")
+)
+
+// Writer writes messages to and AWS SQS queue.
```
[the remark this answers is not in the document]
Comment (danfaizer) [reply storage/s3/writer.go]: Changed the text in [bfe293b](https://github.com/adevinta/vulcan-agent/pull/92/commits/bfe293b89d2a4998909920e0f9939708af6ecab3).
Thanks!

File: storage/s3/writer.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+		cfg: cfg,
+		l:   l,
+	}, nil
+}
+
+// Upload uploads the provided byte array data as a file to an S3 bucket.
+func (w *Writer) UploadCheckData(checkID, kind string, startedAt time.Time, content []byte) (string, error) {
+	st := time.Now()
+	// see http://docs.aws.amazon.com/athena/latest/ug/partitions.html
+	dt := startedAt.Format("dt=2006-01-02")
```
[the remark this answers is not in the document]
Comment (danfaizer) [reply storage/s3/writer.go]: Absolutely right. Moved inside the `if` in [bfe293b](https://github.com/adevinta/vulcan-agent/pull/92/commits/bfe293b89d2a4998909920e0f9939708af6ecab3).
In fact, probably in future refactors that block may be deleted.

File: storage/s3/writer_test.go (hunk 3 above)
Comment (danfaizer) [reply storage/s3/writer_test.go]: Thanks for reporting the typo. Resolved in [bfe293b](https://github.com/adevinta/vulcan-agent/pull/92/commits/bfe293b89d2a4998909920e0f9939708af6ecab3).

File: storage/s3/writer_test.go (hunk 4 above)
Comment (danfaizer) [reply storage/s3/writer_test.go]: Same as above.

File: storage/s3/writer_test.go (hunk 5 above)
Comment (danfaizer) [reply storage/s3/writer_test.go]: Same as above.

File: storage/s3/writer_test.go (hunk 6 above)
Comment (danfaizer) [reply storage/s3/writer_test.go]: Same as above.

File: storage/s3/writer.go (hunk 2 above)
Comment (julianvilas) [reply storage/s3/writer.go:97]: Gotcha, thx! :)

**Review by julianvilas — APPROVED** [review]

**Pull request merged**